"case" function literal incorrectly desugared

```scala
object Test {
  def call(k: (Int, Int) => Unit): Unit = ???
  def test = call({ case (x, y) => ()})
}
```
This fails with:
```scala
funlit.scala:9: error: wrong number of parameters, expected: 2
  def test = call({ case (x, y) => ()})
                  ^
result of funlit.scala after frontend:
package <empty> {
  final lazy module val Test: Test$ = new Test$()
  final module class Test$() extends Object() { this: Test.type => 
    def call(k: Function2[Int, Int, Unit]): Unit = ???
    def test: Unit = 
      Test.call({
        def $anonfun(x$1: <error>): Unit = 
          x$1 match {
            case Tuple2.unapply[Nothing', Nothing'](x @ _, y @ _) => 
              ()
          }
        closure($anonfun)
      })
  }
}
two errors found
```
Because `{ case (x, y) => ()}` is expanded to `x$1 => x$1 match { case (x, y) => ()}`, instead it should be expanded to `(x$1, x$2) => (x$1, x$2) match { case (x, y) => ()}` (like in Scala 2) because this compiles fine in Dotty:
```scala
object Test {
  def call(k: (Int, Int) => Unit): Unit = ???
  def test = call((a, b) => (a, b) match { case (x, y) => ()})
}
```

That's certainly not as specified. I don't know what scalac does here. But
note that the following also compiles in scalac and dotty:

val xs = List(1, 2, 3).zipWithIndex

xs.map{ (case (x, y) => x + y }

And the translation is x$1: (Int, Int) => x$1 match ...

So it seems scalac does some conversion according to the expected type. I
think doing some conversion here is OK, but it should be done for the
version without case. I.e.

     xs.map((x, y) => x + y)

should work. But the example you gave should not work. Can you file another
issue along these lines?



On Tue, Oct 27, 2015 at 3:12 AM, Guillaume Martres <notifications@github.com
> wrote:

> object Test {
>   def call(k: (Int, Int) => Unit): Unit = ???
>   def test = call({ case (x, y) => ()})
> }
>
> This fails with:
>
> funlit.scala:9: error: wrong number of parameters, expected: 2
>   def test = call({ case (x, y) => ()})
>                   ^
> result of funlit.scala after frontend:
> package <empty> {
>   final lazy module val Test: Test$ = new Test$()
>   final module class Test$() extends Object() { this: Test.type =>
>     def call(k: Function2[Int, Int, Unit]): Unit = ???
>     def test: Unit =
>       Test.call({
>         def $anonfun(x$1: <error>): Unit =
>           x$1 match {
>             case Tuple2.unapply[Nothing', Nothing'](x @ _, y @ _) =>
>               ()
>           }
>         closure($anonfun)
>       })
>   }
> }
> two errors found
>
> Because { case (x, y) => ()} is expanded to x$1 => x$1 match { case (x,
> y) => ()}, instead it should be expanded to (x$1, x$2) => (x$1, x$2)
> match { case (x, y) => ()} (like in Scala 2) because this compiles fine
> in Dotty:
>
> object Test {
>   def call(k: (Int, Int) => Unit): Unit = ???
>   def test = call((a, b) => (a, b) match { case (x, y) => ()})
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lampepfl/dotty/issues/873>.
>



-- 
Martin Odersky
EPFL

--- a/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/src/dotty/tools/dotc/ast/Desugar.scala
@@ -556,14 +556,20 @@ object desugar {
       DefDef(nme.ANON_FUN, Nil, params :: Nil, tpt, body).withMods(synthetic),
       Closure(Nil, Ident(nme.ANON_FUN), EmptyTree))
 
-  /** Expand partial function
+  /** If `nparams` == 1, expand partial function
+   *
    *       { cases }
    *  ==>
-   *       x$0 => x$0 match { cases }
+   *       x$1 => x$1 match { cases }
+   *
+   *  If `nparams` != 1, expand instead to
+   *
+   *       (x$0, ..., x${n-1}) => (x$0, ..., x${n-1}) match { cases }
    */
-  def makeCaseLambda(cases: List[CaseDef])(implicit ctx: Context) = {
-    val param = makeSyntheticParameter()
-    Function(param :: Nil, Match(Ident(param.name), cases))
+  def makeCaseLambda(cases: List[CaseDef], nparams: Int = 1)(implicit ctx: Context) = {
+    val params = (1 to nparams).toList.map(makeSyntheticParameter(_))
+    val selector = makeTuple(params.map(p => Ident(p.name)))
+    Function(params, Match(selector, cases))
   }
 
   /** Add annotation with class `cls` to tree:

--- a/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/src/dotty/tools/dotc/ast/Desugar.scala
@@ -562,12 +562,12 @@ object desugar {
    *  ==>
    *       x$1 => x$1 match { cases }
    *
-   *  If `nparams` > 1, expand instead to
+   *  If `nparams` != 1, expand instead to
    *
    *       (x$1, ..., x$n) => (x$0, ..., x${n-1}) match { cases }
    */
   def makeCaseLambda(cases: List[CaseDef], nparams: Int = 1)(implicit ctx: Context) = {
-    val params = (1 to (nparams min 1)).toList.map(makeSyntheticParameter(_))
+    val params = (1 to nparams).toList.map(makeSyntheticParameter(_))
     val selector = makeTuple(params.map(p => Ident(p.name)))
     Function(params, Match(selector, cases))
   }

--- a/src/dotty/tools/dotc/ast/Desugar.scala
+++ b/src/dotty/tools/dotc/ast/Desugar.scala
@@ -562,12 +562,12 @@ object desugar {
    *  ==>
    *       x$1 => x$1 match { cases }
    *
-   *  If `nparams` != 1, expand instead to
+   *  If `nparams` > 1, expand instead to
    *
-   *       (x$0, ..., x${n-1}) => (x$0, ..., x${n-1}) match { cases }
+   *       (x$1, ..., x$n) => (x$0, ..., x${n-1}) match { cases }
    */
   def makeCaseLambda(cases: List[CaseDef], nparams: Int = 1)(implicit ctx: Context) = {
-    val params = (1 to nparams).toList.map(makeSyntheticParameter(_))
+    val params = (1 to (nparams min 1)).toList.map(makeSyntheticParameter(_))
     val selector = makeTuple(params.map(p => Ident(p.name)))
     Function(params, Match(selector, cases))
   }

--- a/src/dotty/tools/dotc/typer/Typer.scala
+++ b/src/dotty/tools/dotc/typer/Typer.scala
@@ -502,6 +502,16 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
     assignType(cpy.If(tree)(cond1, thenp2, elsep2), thenp2, elsep2)
   }
 
+  private def decomposeProtoFunction(pt: Type, defaultArity: Int)(implicit ctx: Context): (List[Type], Type) = pt match {
+    case _ if defn.isFunctionType(pt) =>
+      (pt.dealias.argInfos.init, pt.dealias.argInfos.last)
+    case SAMType(meth) =>
+      val mt @ MethodType(_, paramTypes) = meth.info
+      (paramTypes, mt.resultType)
+    case _ =>
+      (List.range(0, defaultArity) map alwaysWildcardType, WildcardType)
+  }
+
   def typedFunction(tree: untpd.Function, pt: Type)(implicit ctx: Context) = track("typedFunction") {
     val untpd.Function(args, body) = tree
     if (ctx.mode is Mode.Type)
@@ -509,15 +519,7 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
         untpd.TypeTree(defn.FunctionClass(args.length).typeRef), args :+ body), pt)
     else {
       val params = args.asInstanceOf[List[untpd.ValDef]]
-      val (protoFormals, protoResult): (List[Type], Type) = pt match {
-        case _ if defn.isFunctionType(pt) =>
-          (pt.dealias.argInfos.init, pt.dealias.argInfos.last)
-        case SAMType(meth) =>
-          val mt @ MethodType(_, paramTypes) = meth.info
-          (paramTypes, mt.resultType)
-        case _ =>
-          (params map alwaysWildcardType, WildcardType)
-      }
+      val (protoFormals, protoResult) = decomposeProtoFunction(pt, params.length)
 
       def refersTo(arg: untpd.Tree, param: untpd.ValDef): Boolean = arg match {
         case Ident(name) => name == param.name
@@ -629,11 +631,8 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
   def typedMatch(tree: untpd.Match, pt: Type)(implicit ctx: Context) = track("typedMatch") {
     tree.selector match {
       case EmptyTree =>
-        val arity = pt match {
-          case defn.FunctionType(args, _) => args.length
-          case _ => 1
-        }
-        typed(desugar.makeCaseLambda(tree.cases, arity) withPos tree.pos, pt)
+        val (protoFormals, _) = decomposeProtoFunction(pt, 1)
+        typed(desugar.makeCaseLambda(tree.cases, protoFormals.length) withPos tree.pos, pt)
       case _ =>
         val sel1 = typedExpr(tree.selector)
         val selType = widenForMatchSelector(

--- a/src/dotty/tools/dotc/typer/Typer.scala
+++ b/src/dotty/tools/dotc/typer/Typer.scala
@@ -629,7 +629,11 @@ class Typer extends Namer with TypeAssigner with Applications with Implicits wit
   def typedMatch(tree: untpd.Match, pt: Type)(implicit ctx: Context) = track("typedMatch") {
     tree.selector match {
       case EmptyTree =>
-        typed(desugar.makeCaseLambda(tree.cases) withPos tree.pos, pt)
+        val arity = pt match {
+          case defn.FunctionType(args, _) => args.length
+          case _ => 1
+        }
+        typed(desugar.makeCaseLambda(tree.cases, arity) withPos tree.pos, pt)
       case _ =>
         val sel1 = typedExpr(tree.selector)
         val selType = widenForMatchSelector(

--- a/tests/pos/i873.scala
+++ b/tests/pos/i873.scala
@@ -1,0 +1,4 @@
+object Test {
+  def call(k: (Int, Int) => Unit): Unit = ???
+  def test = call({ case (x, y) => ()})
+}

--- a/tests/pos/i873.scala
+++ b/tests/pos/i873.scala
@@ -1,4 +1,10 @@
 object Test {
   def call(k: (Int, Int) => Unit): Unit = ???
   def test = call({ case (x, y) => ()})
+
+  trait X extends Function1[Int, String]
+  implicit def f2x(f: Function1[Int, String]): X = ???
+  ({case _ if "".isEmpty => ""} : X) // allowed, implicit view used to adapt
+
+  // ({case _ if "".isEmpty => 0} : X) // expected String, found Int
 }

